diff --git a/plat/nxp/s32/s32_bl2_el3.c b/plat/nxp/s32/s32_bl2_el3.c
index f80b199b0..82992b026 100644
--- a/plat/nxp/s32/s32_bl2_el3.c
+++ b/plat/nxp/s32/s32_bl2_el3.c
@@ -392,6 +392,31 @@ static void get_fip_images_ranges(uintptr_t *images_base, size_t *images_size)
 	*images_size = MMU_ROUND_UP_TO_PAGE(fip_end - *images_base);
 }
 
+#ifdef HSE_SUPPORT
+static size_t get_fip_size(void)
+{
+   static const uuid_t uuid_null = { {0} };
+   uintptr_t fip_hdr_start, fip_hdr_end;
+   fip_toc_header_t *toc_header;
+   fip_toc_entry_t *toc_entry;
+
+   fip_hdr_start = get_fip_hdr_base();
+   fip_hdr_end = fip_hdr_start + fip_hdr_size;
+
+   toc_header = (fip_toc_header_t *)fip_hdr_start;
+   toc_entry = (fip_toc_entry_t *)(toc_header + 1);
+
+   while ((uintptr_t)toc_entry < fip_hdr_end) {
+       if (!compare_uuids(&toc_entry->uuid, &uuid_null))
+           break;
+
+       toc_entry++;
+   }
+
+   return (size_t)toc_entry->offset_address;
+}
+#endif
+
 int s32_el3_mmu_fixup(const struct s32_mmu_filter *filters, size_t n_filters)
 {
 	const unsigned long code_start = BL_CODE_BASE;
@@ -429,7 +454,12 @@ int s32_el3_mmu_fixup(const struct s32_mmu_filter *filters, size_t n_filters)
 		{
 			.base_pa = get_fip_hdr_page(),
 			.base_va = get_fip_hdr_page(),
+
+#ifdef HSE_SUPPORT
+            .size = MMU_ROUND_UP_TO_PAGE(get_fip_size()),
+#else
 			.size = BL2_BASE - get_fip_hdr_page(),
+#endif
 			.attr = MT_RO | MT_MEMORY | MT_SECURE,
 		},
 		/* Additional images in case the FIP is loaded into memory */
diff --git a/plat/nxp/s32/s32_common.mk b/plat/nxp/s32/s32_common.mk
index e917f1bda..506ae8abf 100644
--- a/plat/nxp/s32/s32_common.mk
+++ b/plat/nxp/s32/s32_common.mk
@@ -400,6 +400,13 @@ ${BL2_W_DTB}: bl2 dtbs ${DTB_SIZE_FILE} ${BL2_PADDING}
 	@${DD} if=${BL2_BIN} of=$@ seek=$$(printf "%d" ${DTB_SIZE}) oflag=seek_bytes
 	@${DD} if=${BL2_PADDING} conv=notrunc >> $@
 
+ifneq (${HSE_SUPPORT},)
+    ${Q}PADDINGHEX=$$($(call hexfilesize,${BUILD_PLAT}/bl2.bin)); \
+    PADDING=$$(printf "%d" $${PADDINGHEX}); \
+    SEEKSIZE=$$(echo "$$(printf '%d' ${DTB_SIZE}) + $${PADDING}" | bc); \
+    dd if=/dev/zero of=$@ seek=$$SEEKSIZE bs=1 count=$$PADDING
+endif
+
 ${BOOT_INFO_SRC}: ${FIP_MMC_OFFSET_FILE} ${FIP_EMMC_OFFSET_FILE} ${FIP_QSPI_OFFSET_FILE} ${FIP_MEMORY_OFFSET_FILE} ${FIP_HDR_SIZE_FILE} ${DTB_SIZE_FILE}
 	${ECHO} "  CREATE  $@"
 	${Q}echo "const unsigned long fip_mmc_offset = $$(cat ${FIP_MMC_OFFSET_FILE});" > ${BOOT_INFO_SRC}
@@ -416,10 +423,16 @@ ${BL2_W_DTB_SIZE_FILE}: ${BL2_W_DTB}
 ${MKIMAGE_FIP_CONF_FILE}: ${BL2_W_DTB_SIZE_FILE} ${FIP_HDR_SIZE_FILE} add_to_fip FORCE
 	${ECHO} "  CREATE  $@"
 	${Q}cp -f ${MKIMAGE_CFG} $@
+ifneq (${HSE_SUPPORT},)
+	${Q}ACTUAL_FIP_SIZE=$$($(call hexfilesize,${BUILD_PLAT}/${FIP_NAME})); \
+	T_SIZE=$$(printf "0x%x" $${ACTUAL_FIP_SIZE}); \
+	echo "DATA_FILE SIZE $$T_SIZE" >> $@
+else
 	${Q}BL2_W_DTB_SIZE=$$(cat ${BL2_W_DTB_SIZE_FILE}); \
 	HDR_SIZE=$$(cat ${FIP_HDR_SIZE_FILE}); \
 	T_SIZE=0x$$($(call hexbc, $${BL2_W_DTB_SIZE}, +, $${HDR_SIZE})); \
 	echo "DATA_FILE SIZE $$T_SIZE" >> $@
+endif
 
 FIP_ALIGN := 16
 all: add_to_fip
